File: app/src/main/java/com/example/bghub/di/MainModule.java
package com.example.bghub.di;

import com.example.bghub.views.main.MainActivity;
import com.example.bghub.views.main.MainContract;
import com.example.bghub.views.main.MainPresenter;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;

/**
 * This is a Dagger module. We use this to auto create the MainSubComponent and bind
 * the {@link MainPresenter} to the graph
 */
@Module
public abstract class MainModule {

    @ActivityScoped
    @Binds
    abstract MainContract.Presenter mainPresenter(MainPresenter presenter);

    @Provides
    static MainContract.View provideView(MainActivity activity) {
        return activity;
    }

}

File: app/src/main/java/com/example/bghub/di/MainModule.kt
package com.example.bghub.di

import android.app.Activity
import com.example.bghub.views.activities.main.MainActivity
import com.example.bghub.views.activities.main.MainContract
import com.example.bghub.views.activities.main.MainContract.Presenter
import com.example.bghub.views.activities.main.MainPresenter
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ActivityComponent
import dagger.hilt.components.SingletonComponent

@InstallIn(ActivityComponent::class)
@Module
abstract class MainModule {

    @Binds
    abstract fun bindActivity(activity: MainActivity): MainContract.View

    @Binds
    abstract fun bindPresenter(impl: MainPresenter): MainContract.Presenter

}

@InstallIn(ActivityComponent::class)
@Module
object MainActivityModule {

    @Provides
    fun bindActivity(activity: Activity): MainActivity {
        return activity as MainActivity
    }
}